R/setkey.R

```diff
@@ -204,49 +204,41 @@ forderv = function(x, by=seq_along(x), retGrp=FALSE, sort=TRUE, order=1L, na.las
-        # that 'arguments are not of same lengths'. But with(DT, order(list(x), list(y))) will return 1L, which is very strange. On top of that, with(DT,
-        # order(x, as.list(10:1)) would return 'unimplemented type list'. It's all very inconsistent. But we HAVE to be consistent with base HERE.
-        if (!as.character(v[[1L]]) %chin% c("+", "-")) break   # FIX for bug #5583
-        if (v[[1L]] == "-") order[i] = -order[i]
-        v = v[[-1L]]
-      }
-      if (is.name(v)) {
-        ix = chmatch(as.character(v), xcols, nomatch=0L)
-        if (ix != 0L) ans = point(ans, i, x, ix) # see 'point' in data.table.R and C-version pointWrapper in assign.c - avoid copies
-        else {
-          v = as.call(list(as.name("list"), v))
-          ans = point(ans, i, eval(v, x, parent.frame()), 1L)
-        }
-      } else {
-        if (!is.object(eval(v, x, parent.frame()))) {
-          v   = as.call(list(as.name("list"), v))
-          ans = point(ans, i, eval(v, x, parent.frame()), 1L) # eval has to make a copy here (not due to list(.), but due to ex: "4-5*y"), unavoidable.
-        } else ans = point(ans, i, list(unlist(eval(v, x, parent.frame()))), 1L)
-      } # else stop("Column arguments to order by in 'forder' should be of type name/symbol (ex: quote(x)) or call (ex: quote(-x), quote(x+5*y))")
+  sub = substitute(list(...))
+  tt = sapply(sub, function(x) is.null(x) || (is.symbol(x) && !nzchar(x)))
+  if (any(tt)) sub[tt] = NULL  # remove any NULL or empty arguments; e.g. test 1962.052: forder(DT, NULL) and forder(DT, )
+  if (length(sub)<2L) return(NULL)  # forder() with no arguments returns NULL consistent with base::order
+  asc = rep.int(1L, length(sub)-1L)  # ascending (1) or descending (-1) per column
+  # the idea here is to intercept - before vectors so that :
+  # i) we can pass the decreasing argument for that column through to C to avoid what normally happens in R (allocate a new vector and apply - to every element first)
+  # ii) - on character vector works; ordinarily in R that fails with type error
+  # We intercept the unevaluated expressions and massage them before evaluating in with(DT) scope or not depending on the first item.
+  for (i in seq.int(2L, length(sub))) {
+    v = sub[[i]]
+    while (is.call(v) && length(v)==2L && ((s<-v[[1L]])=="-" || s=="+")) {
+      if (s=="-") asc[i-1L] = -asc[i-1L]
+      sub[[i]] = v = v[[-1L]]  # remove the leading - (most commonly, or possibly +, --, +-- etc) to avoid the R level allocation upon eval
     }
   }
-  cols = seq_along(ans)
-  # Supported column types are checked at C level
-  o = forderv(ans, cols, sort=TRUE, retGrp=FALSE, order= if (decreasing) -order else order, na.last)
-  if (!length(o)) o = seq_along(ans[[1L]]) else o
+  x = eval(sub[[2L]], parent.frame(), parent.frame())
+  if (is.list(x)) {
+    if (length(x)==0L && is.data.frame(x)) stop("Attempting to order a 0-column data.table or data.frame.")
+    sub[2L] = NULL  # change list(DT, ...) to list(...)
```

Comment:
Ultimately it make sense to redirect users to use `forderv` directly when they want to pass DT into forder. If you agree, then maybe a verbose print for now informing about that? because any change here would be significant (and unnecessary) breaking change. Better to provide a verbose (or eventually non-verbose) print to user, so no one will be harmed.

Reply:
I don't quite follow. There is no breaking change here so far, at least, afaik. This line that changes (`list(DT,...)` to `list(...)`) is basically just maintaining the old usage of forder(DT, -colA, colB) so that still works. That form isn't what forderv is for, iiuc.

Reply:
yes, forder has different API but IMO it worth to make distinction, even if not a breaking change, then just suggests to users. So `forder` is a `order` replacement and `forderv` is meant to be used by passing data.table to it.

Reply:
We can suggest `forderv` in verbose mode I guess. But although `forderv` is meant for passing `DT` to it, a second aspect of `forderv` is that it doesn't accept expressions of columns. I was hoping that users can just use `DT[order(...), ]` as they have been and not need to know that `forder` and `forderv` exist. The optimization should hide these things from them and the news item (15) is consistent with that (e.g. it doesn't mention forder vs forderv). If verbose mode for `DT[order(...), ]` starts to suggest they use `forderv` it might be a bit confusing because i) they didn't even know they were using forder, and ii) neither `forder` or `forderv` are exported or documented.  Hopefully we never need to export them since optimization can take care of it.

Reply:
is `DT[order(...), ]` optimised to `forder(DT, ...)`? if not, then the verbose message wouldn't be printed then. But agree that would be confusing till forder is not exported. Ultimately we want to export it https://github.com/Rdatatable/data.table/issues/3447 "so it can be legally used outside of `[.data.table`"

Reply:
I see now, so verbose message would just be when first argument is `DT`.  `DT[order(...),]` is optimized to `DT[forder(...),]` so wouldn't print even in verbose mode.  I added link to #3447 back to this discussion so we can come back to it.